org settings: Add edit_topics_by_admins_only.

sarahcstringer opened the pull request:
Add a realm setting that allows users to edit topics of any message.

I wanted to check in and see if the wording on the frontend pieces is correct -- the way I edited settings_org to display the success messages seems wordy.

I am going to continue with a few other pieces. In particular, I'll look through the tests to see what else I can add, and I'll try to add a function that updates the UI, similar to ``toggle_allow_message_editing_pencil``.

Fixes #5947

sarahcstringer commented on static/js/ui_init.js:
```diff
@@ -39,7 +39,7 @@ function message_hover(message_row) {
     }
 
-    // But the message edit hover icon is determined by whether the message is still editablex
+    // But the message edit hover icon is determined by whether the message is still editable
```
This doesn't have to do with my code; just found a small typo

gnprice replied:
Well spotted! Would you send this as a separate commit?

For splitting up a commit into separate pieces, `git add -p` is a handy tool.

gnprice commented:
Thanks @sarahcstringer for writing this!

I agree with you that the new text in settings_org feels wordy. The message editing section in the settings UI now feels complicated to me, too: 
![image](https://user-images.githubusercontent.com/28173/30831440-522c0b30-a1fc-11e7-81b9-b51eaa6affed.png)
and I suspect those are related.

Here's an attempt at making the settings UI clearer:
> * Users can edit their messages
>   * Users can edit topics on any message
>   * (time limit)

I think flipping the sense of the option to match "Users can edit their messages" makes it clearer.

Also, thinking about this feature, I think we want a time limit to apply to editing others' topics, just like for one's own messages -- otherwise, in many orgs there will be people for whom the temptation for mischief is just too great. In my sketch above, that'd be the same time limit as for message content, but we could also make it different (maybe 60min by default?):
> * Users can edit their messages
>   * (time limit)
> * Users can edit topics on any message
>   * (time limit)

That feels a little more complex, but I think the right values are probably different for the two time limits.

@rishig, what do you think of the above?

rishig commented:
hm, I hadn't thought about the temptation for mischief. I guess a time limit seems reasonable. I think we'd want admins to be able to edit forever (which is similar to what they can do now), and we'd still want to support a user simultaneously editing every message with a given topic, though they'd have to send a new message to that topic to reset the clock. 

It would be nice to easily support editing the topics from the left sidebar itself, which this makes a bit harder. But I think that's probably fine.

Maybe 24 hours by default for the topic editing?

gnprice commented:
Yeah, admins should be able to edit forever, like now.

I think older messages from beyond the time limit should be out of reach
for a user to edit the topics. One form of obnoxiousness here would be to
reply to some other people's messages on a topic, and go edit the topic to
something that reflects your own idiosyncratic view of the matter, or makes
their messages look foolish. I think it's key that when someone edits your
message's topic, it should at least be recent enough that when you look at
the thread you'll remember what it originally was and know that they
changed it. So thinking through that, it does argue for a long enough time
limit to cover a solid discussion, longer than an hour.

24 hours could work for that, though I wouldn't want to go much beyond it;
I think I'd default to 12 hours. Past 12 or 24 hours, I think the
educational value in a Zulip-savvy user editing someone else's topics to
make the stream demonstrate good use of topics also rapidly diminishes, as
there are new conversations to set and read the topics on instead.

rishig commented:
For 12 vs 24, in general I think it's better to set these defaults with newer realms in mind, since a realm getting even 10s of messages a day probably has an admin that can set the value appropriately for the realm.

For editing topics older messages, I guess the question is what should the last of the following three options do?
![topics-screenshot](https://user-images.githubusercontent.com/890911/30834378-beec80f0-a207-11e7-9888-505dc229d668.png)

gnprice commented:
Hmm. Gosh, the text of those options is confusing -- the "to this topic"
could mean either "messages [sent] to this topic", or "change ... to this
topic", and I'm reading it both ways or alternately, like a Necker cube. :-)

I think a fine solution would be to cut out the third of these options.
Then the two choices would be "this message" and "this and following", and
the latter option is enough to avoid any tedious repetitive clicking --
just go to the first message you want to change. That's also already the
behavior we silently do when an admin edits a topic.

timabbott commented:
One important detail is that the "all messages to topic" one edits all messages with that topic in that stream's history, I think using the backend, and thus is more robust than hitting the home key repeatedly to find the first one.  That feature is certainly useless if we expect folks to restrict topic edits temporally.  We should look at the commit history for when it was added; I vaguely recall it was by user request like 4 years ago :).

gnprice commented:
Setting defaults for newer realms is a good point -- though I think the
quantitative threshold is different. There are a lot of org options, and it
takes a pretty energetic admin to actually go through them all and make
real choices. I'd guess maybe at 300+ messages/day, or like 10+ really
active users, that starts being a good assumption.
​

timabbott commented:
I actually would not at all assume that existing organizations will find out about new settings unless they are advertised somehow.  I've told the org admins of large organizations like RC about new settings 6 months after they were implemented, and seen them be excited about the new setting.

gnprice commented:
Yep, also a good point.

sarahcstringer commented:
Thanks for the discussion! To clarify for moving forward:
- Update the wording for this setting
- Add a time limit for topic edits by users
    - Default is 12 hrs?
    - Is there a max time?
- I need to add the dropdown box when editing older message topics, to select which ones to change; I'm not sure that displays currently in this PR.

timabbott commented:
On the from of the dropdown box for what to change, @rishig @gnprice I think we should try to make this PR be just about the new setting, and not about changes to the "Edit a topic" UI.  While this adding this setting may make us want to update that topic editing UI, I don't see a reason why the changes are actually related, and it's way easier to do 2 things separately than together...

rishig commented:
yup, no need to change the dropdown for this PR, other than to perhaps remove that third option. 

That being said, if there is some intermediate thing that can be merged (e.g. something that allows topics to be edited forever, if at all), it would probably be worth getting that mergeable now, so that we have it for gsoc. It shouldn't affect any existing realms, since the setting will be off by default.

sarahcstringer commented:
Okay, I updated a few pieces of this PR but am working on a new one that adds the time limit.

timabbott commented:
Can you explain what you mean by "adds the time limit"?

sarahcstringer commented:
Sorry, yes -- I was thinking the next step is adding the organization setting for a "topic editability" time limit for users, so that admins could choose to make topics editable up to a certain time after messages are created.

gnprice commented on static/js/settings_org.js:
```diff
@@ -384,26 +390,41 @@ function _set_up() {
                             ui_report.success(
-                                i18n.t("Users can now edit topics for all their messages, and the content of messages which are less than __num_minutes__ minutes old.",
-                                       {num_minutes : data_message_content_edit_limit_minutes}),
+                                i18n.t(success_msg,
```
This won't work for i18n -- strings to be translated have to be string literals.

rishig replied:
iirc we're deleting all these success strings anyway, though maybe confirm with tabbott.

timabbott replied:
We plan to simplify the strings story for our settings, but no work has been done and it's not a small project, so we're stuck with doing something reasonable here for now.

gnprice commented on zerver/views/realm.py:
```diff
@@ -79,16 +80,23 @@ def update_realm(request, user_profile, name=REQ(validator=check_string, default
+        (message_content_edit_limit_seconds is not None and
+            realm.message_content_edit_limit_seconds != message_content_edit_limit_seconds) or
+        (edit_topics_by_admins_only is not None and
+            realm.edit_topics_by_admins_only != edit_topics_by_admins_only)):
```
So, the more I look at this bit of code, and especially at the messages to the user, the more I think that it'd be better to separate these settings at the events level, and just use your handy `do_set_realm_property` mechanism for each of these settings separately, just like we do for most settings.

Some of these settings won't have any effect as long as `allow_message_editing` is False, but I think that's fine. Already `invite_required` and `invite_by_admins_only` work the same way.

This code will get a lot simpler that way, and I think the messages will too.

sarahcstringer commented:
Thanks so much for this feedback @gnprice! I'll get started on updating this :)

timabbott commented on static/js/message_edit.js:
```diff
@@ -40,13 +40,13 @@ function get_editability(message, edit_limit_seconds_buffer) {
-    if (page_params.realm_message_content_edit_limit_seconds + edit_limit_seconds_buffer +
-        now.diffSeconds(message.timestamp * 1000) > 0) {
+    if ((page_params.realm_message_content_edit_limit_seconds + edit_limit_seconds_buffer +
+        now.diffSeconds(message.timestamp * 1000) > 0) && message.sent_by_me) {
```
While this is probably correct, it'd be nice if this was a bit easier to reason about that doesn't feel like we're adding checks on this field in 3 places :)

timabbott commented on static/templates/single_message.handlebars:
```diff
@@ -42,11 +42,14 @@
+                            {{/unless}}
+                            <div class="reaction_button">
+                                <i class="icon-vector-smile" title="{{#tr this}}Add emoji reaction{{/tr}} (:)"></i>
+                            </div>
```
Can you explain why this change is correct?

timabbott commented on zerver/models.py:
```diff
@@ -157,6 +157,7 @@ class Realm(ModelReprMixin, models.Model):
     message_content_edit_limit_seconds = models.IntegerField(default=DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS)  # type: int
     message_retention_days = models.IntegerField(null=True)  # type: Optional[int]
     allow_edit_history = models.BooleanField(default=True)  # type: bool
+    edit_topics_by_admins_only = models.BooleanField(default=False)  # type: bool
```
`allow_community_topic_editing` seems like a good name.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -1022,17 +1022,19 @@ def test_change_realm_message_edit_settings(self):
                                                      allow_message_editing,
-                                                     message_content_edit_limit_seconds))
+                                                     message_content_edit_limit_seconds,
+                                                     edit_topics_by_admins_only))
```
Agreed; the 2 settings already there are directly coupled in the implementation, this one isn't, so rpobably best to avoid it complicating this loop.

We probably want also a simple test for the new setting.

sarahcstringer commented:
I updated the migration file to reflect the original issue, which suggested that the default value for this setting be True, but that we should migrate all past realms in as False.

sarahcstringer commented:
I made the following changes:
- Changed the setting name
- Added a max 24-hr limit for all users editing topics
- Decoupled the message editing settings and made them part of the `property_types` framework.
    - Before I update all the frontend tests, I wanted to check in on the success message portion of this -- previously, the messages for "users can edit the topics of their message" and "users can edit the content of messages up to __ min __ old" went together. Should the messages remain this way, or should they all be separated out into separate success messages?

rishig commented:
I think separated success messages is better (and the simpler the better, since I think we're going to get rid of them at some point.)

sarahcstringer commented:
Okay, sounds good. Thanks!

sarahcstringer commented:
Hello! Apologies, I've lost track of what changes I should make on this PR still. I was thinking I should go over the tests some more, but am no longer sure what other major fixes should happen.

I have other code for decoupling the success messages that I was going to submit separately once this feature was finished.

timabbott commented:
@sarahcstringer sorry for the slow reply!  I think just doing a rebase and testing it carefully is probably all I need from you before trying to merge this.

sarahcstringer commented:
I just updated this PR and split the work up into a few different commits instead of one giant one :)

timabbott commented:
@sarahcstringer do you have time to rebase this?

sarahcstringer commented:
@timabbott Looking at it now; apologies, this PR totally fell off my radar.

timabbott commented:
@sarahcstringer oops, did you update this and then not comment?  I'd had no idea, since GitHub doesn't send notifications if all you do is a push.  

There's some merge conflicts with more of the feedback work we're doing on the settings pages.

timabbott commented on zerver/migrations/0143_realm_allow_community_topic_editing.py:
```diff
@@ -0,0 +1,31 @@
+            name='allow_community_topic_editing',
+            field=models.BooleanField(default=True),
+        ),
+        migrations.RunPython(migrate_past_realms_allow_community_topic_editing_false),
```
You should add a `reverse_code` here that's noop (see how our other `RunPython`s do it) so one can undo this migration.

sarahcstringer commented:
@timabbott I just rebased and have been playing around with this feature. One piece that I need to fix is when a user has narrowed to a specific topic, and the someone changes all of those messages to have a different subject name. Currently, two browser errors pop up rather than being able to handle that change.

sarahcstringer commented:
Update: I think the commit I just pushed fixes that browser error issue.

timabbott commented:
The UI commit merge conflicts with the recent rework completed on simplifying the organization settings codebase.  But I'm working on integrating the rest, since they don't require that commit.

timabbott commented on zerver/migrations/0149_realm_allow_community_topic_editing.py:
```diff
@@ -0,0 +1,31 @@
+    realms = Realm.objects.all()
+    for realm in realms:
+        realm.allow_community_topic_editing = False
+        realm.save()
```
This could be written more cleanly as `Realm.objects.update(allow_community_topic_editing=False)`.

But I think I'm going to temporarily flip the default to False so that it's easier to integrate this partially (without a UI to add it, bugs are less important).

timabbott commented on static/js/ui_init.js:
```diff
@@ -34,7 +34,7 @@ function message_hover(message_row) {
     current_message_hover = message_row;
 
-    if (!message.sent_by_me) {
+    if (!message.sent_by_me && !page_params.realm_allow_community_topic_editing) {
```
I think this block needs to actually use `get_editability` so that if it's past the time limit, this doesn't appear.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -87,7 +87,7 @@ function set_topic_edit_properties(group, message) {
     if (message.subject === compose.empty_topic_placeholder()) {
         group.always_visible_topic_edit = true;
-    } else if (page_params.is_admin) {
+    } else if (page_params.is_admin || page_params.realm_allow_community_topic_editing) {
```
I think this block needs to actually use `get_editability` so that if it's past the time limit, this doesn't appear.

timabbott commented on static/js/message_edit.js:
```diff
@@ -40,15 +40,21 @@ function get_editability(message, edit_limit_seconds_buffer) {
+    if (!message.sent_by_me && (
+        86400 + edit_limit_seconds_buffer + now.diffSeconds(message.timestamp * 1000) <= 0)) {
+        return editability_types.NO;
+    }
```
`get_editability` has a pretty good set of node unit tests covering the existing cases.  We should expand it to cover this.

timabbott commented:
@sarahcstringer I merged most of this PR!  A few things remain:
* I switched the default for new organizations to False for now so this remains disabled until we're ready.  We'll want to add a migration to flip that default back to True when we do the final merge.
* We should extend the message_edit node tests to cover the new editability code paths.  `tools/test-js-with-node --coverage` is helpful for seeing what it's doing (noted above).
* Two checks in JavaScript for whether to show an "edit topic" option need to interact properly with the 24-hour time limit; I think the right approach involves using `get_editability` for those, but needs confirmation.  (I squashed the time commit into the other logic one, since the tests failed in between the two).
* I didn't merge the `message_viewport` change, since I think it's the wrong fix; I'll work on a correct fix for it independently.  So you can not worry about that problem.
* You'll need to rebase the "Settings UI" commit.

Thanks for all your hard work on this!  I think the rest of this should be pretty easy to integrate.

timabbott commented:
Closing in favor of #9176, which is merging shortly.  Thanks for all your work on this @sarahcstringer!
